In [ ]:
%matplotlib inline
import requests
from bs4 import BeautifulSoup
import os
import pandas as pd
import time
import re
import seaborn
from lxml import html
import matplotlib

%config InlineBackend.figure_format = 'svg'
from pylab import rcParams
rcParams['figure.figsize'] = 8, 5

Получаем 250 самых топовых фильмов

In [ ]:
url = 'http://www.imdb.com/chart/top'
response = requests.get(url)
soup = BeautifulSoup(response.text, 'lxml')

Парсим страницу с фильмами и получаем грязные данные с


1.   Названием фильма
2.   Рейтингом
3.   Командой 
4.   Ссылкой на основную страницу




In [ ]:
movies = soup.select('td.titleColumn')
links = [a.attrs.get('href') for a in soup.select('td.titleColumn a')]
crew = [a.attrs.get('title') for a in soup.select('td.titleColumn a')]
ratings = [b.attrs.get('data-value') for b in soup.select('td.posterColumn span[name=ir]')]

In [ ]:
data_frame = {
    "movies": movies,
    "ratings": ratings,
    "crew": crew,
    "links": links
}

data = pd.DataFrame(data_frame)
data

movies            ratings  \
0    [\n      1.\n      , [The Shawshank Redemption...  9.239907206438563   
1    [\n      2.\n      , [The Godfather], \n, [(19...  9.160916749679123   
2    [\n      3.\n      , [The Dark Knight], \n, [(...  8.992906485653204   
3    [\n      4.\n      , [The Godfather: Part II],...  8.990104037824768   
4    [\n      5.\n      , [12 Angry Men], \n, [(195...  8.950443397398235   
..                                                 ...                ...   
245  [\n      246.\n      , [Beauty and the Beast],...  8.017242583639977   
246  [\n      247.\n      , [Gandhi], \n, [(1982)],...  8.015655688964149   
247  [\n      248.\n      , [Jai Bhim], \n, [(2021)...   8.01545672669337   
248  [\n      249.\n      , [The Help], \n, [(2011)...  8.014693602269382   
249  [\n      250.\n      , [The Handmaiden], \n, [...  8.012689267557525   

                                                  crew               links  
0    Frank Darabont (dir.), Tim Robbins, Morgan Fre...   /title/tt0111161/  
1    Francis Ford Coppola (dir.), Marlon Brando, Al...   /title/tt0068646/  
2    Christopher Nolan (dir.), Christian Bale, Heat...   /title/tt0468569/  
3    Francis Ford Coppola (dir.), Al Pacino, Robert...   /title/tt0071562/  
4        Sidney Lumet (dir.), Henry Fonda, Lee J. Cobb   /title/tt0050083/  
..                                                 ...                 ...  
245  Gary Trousdale (dir.), Paige O'Hara, Robby Benson   /title/tt0101414/  
246  Richard Attenborough (dir.), Ben Kingsley, Joh...   /title/tt0083987/  
247        T.J. Gnanavel (dir.), Suriya, Lijo Mol Jose  /title/tt15097216/  
248        Tate Taylor (dir.), Emma Stone, Viola Davis   /title/tt1454029/  
249    Park Chan-wook (dir.), Kim Min-hee, Ha Jung-woo   /title/tt4016934/  

[250 rows x 4 columns]

Несколько функциий для преобзразования грязных заголовков с названием фильмов


In [ ]:
def process_text_mv(text):
    data = text.get_text().split("\n")
    return data[2].strip()

In [ ]:
def process_text_year(text):
    data = text.get_text().split("\n")
    return data[3].strip()[1:-1]

In [ ]:
data["title"] = data.movies.apply(process_text_mv)
data["year"] = data.movies.apply(process_text_year)

In [ ]:
data.head()

movies            ratings  \
0  [\n      1.\n      , [The Shawshank Redemption...  9.239907206438563   
1  [\n      2.\n      , [The Godfather], \n, [(19...  9.160916749679123   
2  [\n      3.\n      , [The Dark Knight], \n, [(...  8.992906485653204   
3  [\n      4.\n      , [The Godfather: Part II],...  8.990104037824768   
4  [\n      5.\n      , [12 Angry Men], \n, [(195...  8.950443397398235   

                                                crew              links  \
0  Frank Darabont (dir.), Tim Robbins, Morgan Fre...  /title/tt0111161/   
1  Francis Ford Coppola (dir.), Marlon Brando, Al...  /title/tt0068646/   
2  Christopher Nolan (dir.), Christian Bale, Heat...  /title/tt0468569/   
3  Francis Ford Coppola (dir.), Al Pacino, Robert...  /title/tt0071562/   
4      Sidney Lumet (dir.), Henry Fonda, Lee J. Cobb  /title/tt0050083/   

                      title  year  
0  The Shawshank Redemption  1994  
1             The Godfather  1972  
2           The Dark Knight  2008  
3    The Godfather: Part II  1974  
4              12 Angry Men  1957

In [ ]:
data.crew[1]

'Francis Ford Coppola (dir.), Marlon Brando, Al Pacino'

Задача 1.
Получить список всех участников команд (убрать постфикс `(dir.)` у директоров). Найти количество уникальных имен. 

Задача 2.
Получить список всех директоров. Построить распределение (визуализация обязательна) и найти директора(ов), у которого(ых) больше всего фильмов в топе.

Задача 3. Найти участника(ов) команд, у которого(ых) больше всего фильмов в топе.

Задача 4. Построить распределение топовых фильмов по годам

Задача 5. Построить распределение топовых фильмов рейтингу:


*   Округлить до сотых и отрисовать
*   Округлить до десятых и отрисовать
*   Округлить до целых и отрисовать

Задача 6. Используйте ссылку `http://www.imdb.com/` и динамически добавляя ссылку на фильм из колонки `links` соберите описание фильмов. Пример поиска необходимого текст представлен ниже. Добавьте эту информацию в исходный датафрейм.










In [ ]:
 film_url = 'http://www.imdb.com/title/tt0068646/'
 html_text = requests.get(film_url).text

In [ ]:
soup = BeautifulSoup(html_text, 'lxml')
soup.find_all(attrs={"name": "description"})[0].get("content")

'The Godfather: Directed by Francis Ford Coppola. With Marlon Brando, Al Pacino, James Caan, Richard S. Castellano. The aging patriarch of an organized crime dynasty in postwar New York City transfers control of his clandestine empire to his reluctant youngest son.'


Задача 7. С помощью регулярных выражений найдите в текстах описания фильмов имена актеров, если это возможно (то есть - они там упоминаются). Добавьте эту информацию в датафрейм и посчитайте, для скольких фильмов получилось это сдлеать. Подсказка: перечисление актеров иногда начинается с with. 


Задача 8. Распарсите исходную страницу с топ-250 фильмов и получите для каждого фильмов количество оценок. Пример текста в исходной странице:`9.2 based on 2,554,919 user ratings`. Из этого текста нужно достать = `2,554,919`. Подсказка: используя метод `find_all` с аргументом `attrs` в который нужно передать `{'class': 'ratingColumn imdbRating'}` от страници `http://www.imdb.com/chart/top`. Найдите 10 фильмов (вывести их название и кол-во оценок) у которых, самое большое количество оценок.


Задача 9. Постройте топ 10 слов, которые встречаются 

1.   в названии фильмов
2.   в описании фильмов
3.   в названии и в описании фильмов


Задача 10*. Спарсить теги фильмов со страницы каждого фильма, так например для `The Godfather` ожидается получить список из `Crime, Drama`. (Можно увидеть эти теги по картинкой к фильму)
